Applications/Spire/Source/Ui/TextAreaBox.cpp

```diff
@@ -0,0 +1,630 @@
+      parent) {}
+
+TextAreaBox::TextAreaBox(std::shared_ptr<TextModel> model, QWidget* parent)
+    : QWidget(parent),
+      m_placeholder_styles{[=] { commit_placeholder_style(); }},
+      m_text_edit_styles{[=] { commit_style(); }},
+      m_model(std::move(model)),
+      m_submission(m_model->get_current()) {
+  m_text_edit = new ContentSizedTextEdit(m_model->get_current(), this);
+  m_text_edit->installEventFilter(this);
+  m_stacked_widget = new QStackedWidget(this);
+  m_stacked_widget->addWidget(m_text_edit);
+  setFocusProxy(m_text_edit);
+  connect(m_text_edit->document(), &QTextDocument::contentsChanged, this,
+    &TextAreaBox::on_text_changed);
+  m_placeholder = new ElidedLabel(this);
+  m_placeholder->setFixedSize(0, 0);
+  m_stacked_widget->addWidget(m_placeholder);
+  m_placeholder->setCursor(m_text_edit->cursor());
+  m_placeholder->setAttribute(Qt::WA_TransparentForMouseEvents);
+  m_scroll_box = new ScrollBox(m_stacked_widget, this);
+  m_scroll_box->setFocusProxy(m_text_edit);
+  m_scroll_box->set(ScrollBox::DisplayPolicy::NEVER,
+    ScrollBox::DisplayPolicy::ON_OVERFLOW);
+  m_scroll_box->setSizePolicy(QSizePolicy::Minimum, QSizePolicy::Minimum);
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  layout->addWidget(m_scroll_box);
+  proxy_style(*this, *m_scroll_box);
+  add_pseudo_element(*this, Placeholder());
+  connect_style_signal(*this, [=] { on_style(); });
+  connect_style_signal(*this, Placeholder(), [=] { on_style(); });
+  set_style(*this, DEFAULT_STYLE());
+  connect(m_text_edit, &QTextEdit::cursorPositionChanged, this,
+    &TextAreaBox::on_cursor_position);
+  m_current_connection = m_model->connect_current_signal(
+    [=] (const auto& value) { on_current(value); });
+  connect(m_text_edit->document()->documentLayout(),
+    &QAbstractTextDocumentLayout::documentSizeChanged,
+    [=] (const auto& size) { m_stacked_widget->adjustSize(); });
+}
+
+const std::shared_ptr<TextModel>& TextAreaBox::get_model() const {
+  return m_model;
+}
+
+const QString& TextAreaBox::get_submission() const {
+  return m_submission;
+}
+
+void TextAreaBox::set_placeholder(const QString& value) {
+  m_placeholder_text = value;
+  update_placeholder_text();
+}
+
+bool TextAreaBox::is_read_only() const {
+  return m_text_edit->isReadOnly();
+}
+
+void TextAreaBox::set_read_only(bool read_only) {
+  m_text_edit->setReadOnly(read_only);
+  m_scroll_box->get_vertical_scroll_bar().set_position(0);
+  if(read_only) {
+    m_scroll_box->set_vertical(ScrollBox::DisplayPolicy::NEVER);
+    match(*this, ReadOnly());
+  } else {
+    m_scroll_box->set_vertical(ScrollBox::DisplayPolicy::ON_OVERFLOW);
+    unmatch(*this, ReadOnly());
+  }
+  update_layout();
+}
+
+connection TextAreaBox::connect_submit_signal(
+    const SubmitSignal::slot_type& slot) const {
+  return m_submit_signal.connect(slot);
+}
+
+QSize TextAreaBox::sizeHint() const {
+  return m_scroll_box->sizeHint();
```

Comment:
I think the sizeHint function is not necessary, because the m_scroll_box is placed in the layout of the TextAreaBox and Qt can handle it automatically.